norse/torch/functional/adjoint/lif_refrac_adjoint.py

```diff
@@ -112,6 +113,122 @@ def lif_refrac_adjoint_step(
+        ctx.rho_reset = p.rho_reset
+        ctx.dt = dt
+        s = LIFRefracState(lif_state, rho)
+        z_new, s_new = lif_refrac_step_sparse(
+            input, s, input_weights, recurrent_weights, p, dt
+        )
+
+        # dv before spiking
+        dv_m = p.lif.tau_mem_inv * ((p.lif.v_leak - s_new.lif.v) + s.lif.i)
+        # dev after spiking
+        dv_p = p.lif.tau_mem_inv * ((p.lif.v_leak - s_new.lif.v) + s.lif.i)
+        refrac_period_end = (s.rho == 1).float()
+        ctx.save_for_backward(
+            input,
+            z_new,
+            s_new.rho,
+            refrac_period_end,
+            dv_m.sparse_mask(z_new),
+            dv_p.sparse_mask(z_new),
+            input_weights,
+            recurrent_weights,
+        )
+        return z_new, s_new.lif.v, s_new.lif.i, s_new.rho
+
+    @staticmethod
+    def backward(
+        ctx,
+        doutput: torch.Tensor,
+        lambda_v: torch.Tensor,
+        lambda_i: torch.Tensor,
+        lambda_rho: torch.Tensor,
+    ):
+        (
+            input_tensor,
+            z,
+            refrac_count,
+            refrac_period_end,
+            dv_m,
+            dv_p,
+            input_weights,
+            recurrent_weights,
+        ) = ctx.saved_tensors
+        tau_syn_inv = ctx.tau_syn_inv
+        tau_mem_inv = ctx.taus_mem_inv
+        dt = ctx.dt
+        refrac_mask = heaviside(refrac_count)
+        dv_m = dv_m.to_dense()
+        dv_p = dv_p.to_dense()
+        z = z.to_dense()
+
+        dw_input = lambda_i.t().mm(input_tensor)
+        dw_rec = lambda_i.t().mm(z)
+
+        # lambda_i decay
+        dlambda_i = tau_syn_inv * ((1 - refrac_mask) * lambda_v - lambda_i)
+        lambda_i = lambda_i + dt * dlambda_i
+
+        # lambda_v decay
+        lambda_v = lambda_v - (1 - refrac_mask) * tau_mem_inv * dt * lambda_v
+
+        output_term = z * (1 / dv_m) * (doutput)
+        output_term[output_term != output_term] = 0.0
+
+        jump_term = z * (dv_p / dv_m)
+        jump_term[jump_term != jump_term] = 0.0
+
+        lambda_v = (1 - z) * lambda_v + jump_term * lambda_v + output_term
+        lambda_rho = (
+            1 - refrac_period_end
+        ) * lambda_rho + refrac_period_end * lambda_v * dv_p
+        dinput = lambda_i.mm(input_weights)
+        drecurrent = lambda_i.mm(recurrent_weights)
+        return (
+            dinput,
+            drecurrent,
+            lambda_v,
+            lambda_i,
+            lambda_rho,
+            dw_input,
+            dw_rec,
+            None,
+            None,
+        )
+
+
+def lif_refrac_adjoint_step_sparse(
+    intput: torch.Tensor,
+    s: LIFRefracState,
+    input_weights: torch.Tensor,
+    recurrent_weights: torch.Tensor,
+    p: LIFRefracParameters,
+    dt: float = 0.001,
+) -> Tuple[torch.Tensor, LIFRefracState]:
+    z, v, i, rho = LIFSparseAdjointRefractFunction.apply(
+        input, s.lif, s.lif, s.lif, s.rho, input_weights, recurrent_weights, p, dt
```

Comment:
The `s.lif, s.lif, s.lif` arguments look strange. Shouldn't it be a single `LIFState` instance, as per `LIFSparseAdjointRefractFunction::forward`? This could also be where your type error comes from.

Reply:
Oh, thank you. Obviously a mistake here, for some reason I did not pay attention when I looked through it many times

Reply:
I fixed it but it didn't fix the error.

Reply:
I think the `self` parameter in `lif_refrac` was "pushing" the other parameters down, so the type checker interpreted the refrac parameters as the dt. Let's see if the checks agree!